## Coursera Capstone Project
This notebook will mainly be used for the capstone project for the IBM Data Science Professional Certificate on Coursera.

In [1]:
import pandas as pd
import numpy as np

## Table of contents
* [Introduction - Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)

## Introduction - Business Problem <a name="introduction"></a>
***
When traveling, it is always important to get a hotel in the right area. If you are in a big city that you've never been to, you may not want to try to figure out how to get around whether by taxi, bus, or some other public transportation. You may even just want to avoid spending money on getting around and instead be able to walk everywhere you want to visit. For that reason, we can compare your preferences to the different areas of the city you want to visit and what they have to offer to make a recommendation of where you may want to get a hotel. These preferences could be from a previous trip, so, for example, if you visited Paris last year and want to go to London this year, you could compare neighborhoods that you went to and enjoyed in New York to the neighborhoods in London to see where you might prefer to stay.

## Data <a name="data"></a>
***
The data I will be using will come from FourSquare based on the neighborhoods in London and Paris. The project will be set up so that you can change the two cities, but for the purposes of demonstration that's what I will use. First, I will get the neighborhood latitudes and longitudes which will be needed for gathering the data from FourSquare. I will then use that data to determine similar neighborhoods based on the types of venues that are located in each area. That way I can find which neighborhoods are similar and then make recommendations to the user based on the findings. If the user enjoyed cafes, parks, and museums in Paris, they would receive recommendations on where to stay in London that have similar venues.

### Gathering Neighborhoods

First, I want to get neighborhoods in both cities. This involves using the postal codes for different areas of each city, and then graphing them to see if there is any editing I want to do in terms of how many neighborhoods I want to compare, how far spread out the neighborhoods are, and any other issues that might be noticeable when visualizing the neighborhoods on a map.

#### London
I was able to find information on London and its postcodes on Wikipedia. I will scrape the table with the information and then narrow down the list to neighborhoods at the heart of the city as the neighborhoods for Paris are all pretty central. I want them to be similarly set up and not have one spread out more than the other.

First, I will install the necessary packages to scrape and store the data into a dataframe.

In [2]:
! pip install html-table-parser-python3

import urllib.request

from html_table_parser import HTMLTableParser

Next I get the table and store it in a dataframe. After checking the first five columns, I can see I need to edit the column names, so I do that here as well. Finally, I will not need the Dial code, but I will use the OS Grid ref to get the Latitude and Longitude.

In [3]:
def url_get_contents(url):
    req = urllib.request.Request(url=url)
    f = urllib.request.urlopen(req)
    
    return f.read()

xhtml = url_get_contents('https://en.wikipedia.org/wiki/List_of_areas_of_London').decode('utf-8')

parser = HTMLTableParser()

parser.feed(xhtml)

london_df = pd.DataFrame(parser.tables[1])

london_df.columns = london_df.loc[0]
london_df.drop(0, inplace=True)

london_df = london_df.iloc[:, [0,1,2,3,5]]

london_df

0           Location                      London borough       Post town  \
1         Abbey Wood              Bexley,  Greenwich [7]          LONDON   
2              Acton  Ealing, Hammersmith and Fulham [8]          LONDON   
3          Addington                         Croydon [8]         CROYDON   
4         Addiscombe                         Croydon [8]         CROYDON   
5        Albany Park                              Bexley  BEXLEY, SIDCUP   
..               ...                                 ...             ...   
527         Woolwich                           Greenwich          LONDON   
528   Worcester Park        Sutton, Kingston upon Thames  WORCESTER PARK   
529  Wormwood Scrubs              Hammersmith and Fulham          LONDON   
530          Yeading                          Hillingdon           HAYES   
531         Yiewsley                          Hillingdon    WEST DRAYTON   

0   Postcode district OS grid ref  
1                 SE2    TQ465785  
2              W3, W4    TQ205805  
3                 CR0    TQ375645  
4                 CR0    TQ345665  
5           DA5, DA14    TQ478728  
..                ...         ...  
527              SE18    TQ435795  
528               KT4    TQ225655  
529               W12    TQ225815  
530               UB4    TQ115825  
531               UB7    TQ063804  

[531 rows x 5 columns]

I now want to get the Latitude and Longitude for each location by using the OS grid ref and add those to the dataframe. A small number of neighborhoods did not have OS grid ref, so I just get rid of those rows first, and then I have to reset the indices. Then I iterate over every row and get the latitude and longitude from the OS grid ref and add those to the dataframe.

In [4]:
! pip install OSGridConverter
from OSGridConverter import grid2latlong

london_df['OS grid ref'].replace('', np.nan, inplace=True)
london_df.dropna(subset=['OS grid ref'], inplace=True)
london_df.reset_index(drop=True, inplace=True)
for index, row in london_df.iterrows():
    os_grid_ref = row["OS grid ref"]
    
    lat_long_coords = grid2latlong(os_grid_ref)
    latitude = lat_long_coords.latitude
    longitude = lat_long_coords.longitude

    london_df.loc[london_df.index[index], 'Latitude'] = latitude
    london_df.loc[london_df.index[index], 'Longitude'] = longitude
    
    
london_df

0           Location                      London borough       Post town  \
0         Abbey Wood              Bexley,  Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham [8]          LONDON   
2          Addington                         Croydon [8]         CROYDON   
3         Addiscombe                         Croydon [8]         CROYDON   
4        Albany Park                              Bexley  BEXLEY, SIDCUP   
..               ...                                 ...             ...   
524         Woolwich                           Greenwich          LONDON   
525   Worcester Park        Sutton, Kingston upon Thames  WORCESTER PARK   
526  Wormwood Scrubs              Hammersmith and Fulham          LONDON   
527          Yeading                          Hillingdon           HAYES   
528         Yiewsley                          Hillingdon    WEST DRAYTON   

0   Postcode district OS grid ref   Latitude  Longitude  
0                 SE2    TQ465785  51.486484   0.109318  
1              W3, W4    TQ205805  51.510591  -0.264585  
2                 CR0    TQ375645  51.362934  -0.025780  
3                 CR0    TQ345665  51.381625  -0.068126  
4           DA5, DA14    TQ478728  51.434929   0.125663  
..                ...         ...        ...        ...  
524              SE18    TQ435795  51.496238   0.066504  
525               KT4    TQ225655  51.375352  -0.240950  
526               W12    TQ225815  51.519148  -0.235411  
527               UB4    TQ115825  51.530413  -0.393669  
528               UB7    TQ063804  51.512539  -0.469279  

[529 rows x 7 columns]

I now want to visualize the neighborhoods of London on a map so I can see if I want to eliminate any further rows from my data because the neighborhoods are not in the area I would like. First I install and import the necessary libraries, then center the map on London, and then graph the neighborhoods on the map to determine how spread out they are.

In [5]:
! pip install folium
import folium

! pip install geopy
from geopy.geocoders import Nominatim

     |████████████████████████████████| 94 kB 5.4 MB/s  eta 0:00:01


In [6]:
address = 'London, England'

geolocator = Nominatim(user_agent="london_explorer")
location = geolocator.geocode(address)
london_latitude = location.latitude
london_longitude = location.longitude
print('The geographical coordinates of London are {}, {}.'.format(london_latitude, london_longitude))

The geographical coordinates of London are 51.5073219, -0.1276474.


In [7]:
map_london = folium.Map(location=[london_latitude, london_longitude], zoom_start=10)

for lat, lng, neighborhood in zip(london_df['Latitude'], london_df['Longitude'], london_df['Location']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False
    ).add_to(map_london)

map_london

As we can see from the map above, there is a large amount of neighborhoods, and a large amount of spread for those neighborhoods. I know that Paris is split into much fewer neighborhoods, so I am going to narrow the London Neighborhoods to only those with Post town of London.

In [8]:
london_only_df = london_df[london_df["Post town"] == 'LONDON'].reset_index(drop=True)

london_only_df

0           Location                      London borough Post town  \
0         Abbey Wood              Bexley,  Greenwich [7]    LONDON   
1              Acton  Ealing, Hammersmith and Fulham [8]    LONDON   
2            Aldgate                           City [10]    LONDON   
3            Aldwych                    Westminster [10]    LONDON   
4            Anerley                        Bromley [11]    LONDON   
..               ...                                 ...       ...   
292       Wood Green                            Haringey    LONDON   
293         Woodford                           Redbridge    LONDON   
294    Woodside Park                              Barnet    LONDON   
295         Woolwich                           Greenwich    LONDON   
296  Wormwood Scrubs              Hammersmith and Fulham    LONDON   

0   Postcode district OS grid ref   Latitude  Longitude  
0                 SE2    TQ465785  51.486484   0.109318  
1              W3, W4    TQ205805  51.510591  -0.264585  
2                 EC3    TQ334813  51.514885  -0.078356  
3                 WC2    TQ307810  51.512819  -0.117388  
4                SE20    TQ345695  51.408585  -0.066989  
..                ...         ...        ...        ...  
292               N22    TQ305905  51.598237  -0.116745  
293          IG8, E18    TQ405915  51.604820   0.028068  
294               N12    TQ256925  51.617324  -0.186791  
295              SE18    TQ435795  51.496238   0.066504  
296               W12    TQ225815  51.519148  -0.235411  

[297 rows x 7 columns]

In [9]:
map_london = folium.Map(location=[london_latitude, london_longitude], zoom_start=10)

for lat, lng, neighborhood in zip(london_only_df['Latitude'], london_only_df['Longitude'], london_only_df['Location']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False
    ).add_to(map_london)

map_london

This is better, but there are still 297 neighborhoods and a lot of spread. I will narrow it down further by limiting the distance from the city center.

In [10]:
import math

def within_city_center(lat, long):
    radius = math.sqrt(math.pow(lat - london_latitude, 2) + math.pow(long - london_longitude, 2))
    return radius < .04

In [11]:
for index, row in london_only_df.iterrows():
    london_only_df.loc[london_only_df.index[index], 'City Center'] = within_city_center(row['Latitude'], row['Longitude'])
    
london_only_df

0           Location                      London borough Post town  \
0         Abbey Wood              Bexley,  Greenwich [7]    LONDON   
1              Acton  Ealing, Hammersmith and Fulham [8]    LONDON   
2            Aldgate                           City [10]    LONDON   
3            Aldwych                    Westminster [10]    LONDON   
4            Anerley                        Bromley [11]    LONDON   
..               ...                                 ...       ...   
292       Wood Green                            Haringey    LONDON   
293         Woodford                           Redbridge    LONDON   
294    Woodside Park                              Barnet    LONDON   
295         Woolwich                           Greenwich    LONDON   
296  Wormwood Scrubs              Hammersmith and Fulham    LONDON   

0   Postcode district OS grid ref   Latitude  Longitude City Center  
0                 SE2    TQ465785  51.486484   0.109318       False  
1              W3, W4    TQ205805  51.510591  -0.264585       False  
2                 EC3    TQ334813  51.514885  -0.078356       False  
3                 WC2    TQ307810  51.512819  -0.117388        True  
4                SE20    TQ345695  51.408585  -0.066989       False  
..                ...         ...        ...        ...         ...  
292               N22    TQ305905  51.598237  -0.116745       False  
293          IG8, E18    TQ405915  51.604820   0.028068       False  
294               N12    TQ256925  51.617324  -0.186791       False  
295              SE18    TQ435795  51.496238   0.066504       False  
296               W12    TQ225815  51.519148  -0.235411       False  

[297 rows x 8 columns]

In [12]:
london_city_center_df = london_only_df[london_only_df['City Center'] == True].reset_index(drop=True)

london_city_center_df

0                          Location      London borough Post town  \
0                           Aldwych    Westminster [10]    LONDON   
1                          Bankside      Southwark [14]    LONDON   
2                          Barbican           City [14]    LONDON   
3                         Barnsbury      Islington [17]    LONDON   
4                         Belgravia    Westminster [22]    LONDON   
5                       Blackfriars           City [27]    LONDON   
6                        Bloomsbury         Camden [29]    LONDON   
7                       Camden Town         Camden [40]    LONDON   
8                     Charing Cross         Westminster    LONDON   
9                         Chinatown         Westminster    LONDON   
10                      Clerkenwell           Islington    LONDON   
11                    Covent Garden         Westminster    LONDON   
12              Elephant and Castle           Southwark    LONDON   
13                       Farringdon     Islington, City    LONDON   
14                         Finsbury           Islington    LONDON   
15                        Fitzrovia              Camden    LONDON   
16                          Holborn              Camden    LONDON   
17                       Kennington  Lambeth, Southwark    LONDON   
18                     King's Cross   Camden, Islington    LONDON   
19                    Knightsbridge         Westminster    LONDON   
20                          Lambeth             Lambeth    LONDON   
21  Marylebone (also St Marylebone)         Westminster    LONDON   
22                          Mayfair         Westminster    LONDON   
23                         Millbank         Westminster    LONDON   
24                        Newington           Southwark    LONDON   
25                        Nine Elms          Wandsworth    LONDON   
26                      Pentonville           Islington    LONDON   
27                          Pimlico         Westminster    LONDON   
28                    Primrose Hill              Camden    LONDON   
29                             Soho         Westminster    LONDON   
30                       Somerstown              Camden    LONDON   
31                       St James's         Westminster    LONDON   
32                         St Giles              Camden    LONDON   
33                        St Luke's           Islington    LONDON   
34                       St Pancras              Camden    LONDON   
35                           Temple   City, Westminster    LONDON   
36                         Vauxhall             Lambeth    LONDON   
37                         Walworth           Southwark    LONDON   
38                      Westminster         Westminster    LONDON   

0  Postcode district OS grid ref   Latitude  Longitude City Center  
0                WC2    TQ307810  51.512819  -0.117388        True  
1                SE1    TQ325795  51.498921  -0.092006        True  
2           EC1, EC2    TQ322818  51.519660  -0.095466        True  
3                 N1    TQ305845  51.544318  -0.118974        True  
4                SW1    TQ283792  51.497193  -0.152637        True  
5                EC4    TQ318808  51.510767  -0.101607        True  
6                WC1    TQ299820  51.521990  -0.128550        True  
7                NW1    TQ295845  51.544548  -0.133398        True  
8                WC2    TQ305805  51.508372  -0.120456        True  
9                 W1    TQ297808  51.511252  -0.131875        True  
10               EC1    TQ315825  51.526113  -0.105296        True  
11               WC2    TQ303809  51.512013  -0.123190        True  
12   SE1, SE11, SE17    TQ319789  51.493669  -0.100876        True  
13               EC1    TQ315818  51.519823  -0.105558        True  
14               EC1    TQ315825  51.526113  -0.105296        True  
15                W1    TQ293816  51.518533  -0.137347        True  
16          WC1, WC2    TQ305815  51.517359  -0.120085        True

In [13]:
map_london = folium.Map(location=[london_latitude, london_longitude], zoom_start=12)

for lat, lng, neighborhood in zip(london_city_center_df['Latitude'], london_city_center_df['Longitude'], london_city_center_df['Location']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False
    ).add_to(map_london)

map_london

The above map looks much better for comparing with Paris. The number of neighborhoods is similar, and the location seems better suited to visiting London and not staying too far from the majority of the attractions.

#### Paris
For Paris, I will use the 20 Arrondissements. I will run through the list getting each latitude and longitude and adding them to the dataframe.

In [14]:
arrondissement = ["1st", "2nd", "3rd", "4th", "5th", "6th", "7th", "8th", "9th", "10th", "11th", "12th", "13th", "14th", "15th", "16th", "17th", "18th", "19th", "20th"]
latitudes = []
longitudes = []
for arr in arrondissement:
    address = "{} arrondissement, Paris, France".format(arr)
    location = geolocator.geocode(address)
    latitudes.append(location.latitude)
    longitudes.append(location.longitude)

paris_df = pd.DataFrame(list(zip(arrondissement, latitudes, longitudes)), columns = ["Arrondissement", "Latitude", "Longitude"])

paris_df
    

Arrondissement   Latitude  Longitude
0             1st  48.864614   2.334396
1             2nd  48.868743   2.341688
2             3rd  48.864212   2.360936
3             4th  48.856202   2.355619
4             5th  48.845973   2.344350
5             6th  48.850433   2.332951
6             7th  48.857028   2.320195
7             8th  48.870905   2.312152
8             9th  48.876019   2.339962
9            10th  48.876126   2.359839
10           11th  48.858416   2.379703
11           12th  48.839615   2.395752
12           13th  48.832397   2.355583
13           14th  48.833059   2.326959
14           15th  48.841370   2.300383
15           16th  48.863171   2.275765
16           17th  48.884224   2.322364
17           18th  48.890012   2.346467
18           19th  48.889343   2.384360
19           20th  48.865042   2.398929

In [15]:
address= 'Paris, France'
location = geolocator.geocode(address)
paris_latitude = location.latitude
paris_longitude = location.longitude

map_paris = folium.Map(location=[paris_latitude, paris_longitude], zoom_start=12)

for lat, lng, neighborhood in zip(paris_df['Latitude'], paris_df['Longitude'], paris_df['Arrondissement']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False
    ).add_to(map_paris)

map_paris

Now that I have comparable data for both cities, the next step is to start gathering data that 
can be used to compare the neighborhoods.

### Foursquare

Here, I will gather data on the venues in each neighborhood. I want to gather information on the types of venues and their locations which will help determine clusters later on.

In [16]:
# The code was removed by Watson Studio for sharing.

In [17]:
import requests

In [18]:
def getNearbyVenues(names, latitudes, longitudes, radius=800):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [19]:
paris_venues = getNearbyVenues(names=paris_df['Arrondissement'], latitudes=paris_df['Latitude'], longitudes=paris_df['Longitude'])
paris_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0          1st              48.864614                2.334396   
1          1st              48.864614                2.334396   
2          1st              48.864614                2.334396   
3          1st              48.864614                2.334396   
4          1st              48.864614                2.334396   

                       Venue  Venue Latitude  Venue Longitude  \
0  Musée des Arts Décoratifs       48.863077         2.333393   
1                  Ellsworth       48.865528         2.337057   
2        Place des Pyramides       48.863924         2.332224   
3                Kosyuen 華修園       48.864163         2.333567   
4     Jardin du Palais Royal       48.864941         2.337728   

      Venue Category  
0         Art Museum  
1  French Restaurant  
2              Plaza  
3           Tea Room  
4             Garden

In [20]:
london_venues = getNearbyVenues(names=london_city_center_df['Location'], latitudes=london_city_center_df['Latitude'], longitudes=london_city_center_df['Longitude'])
london_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0      Aldwych              51.512819               -0.117388   
1      Aldwych              51.512819               -0.117388   
2      Aldwych              51.512819               -0.117388   
3      Aldwych              51.512819               -0.117388   
4      Aldwych              51.512819               -0.117388   

             Venue  Venue Latitude  Venue Longitude Venue Category  
0   Lyceum Theatre       51.511598        -0.119785        Theater  
1   Somerset House       51.510786        -0.117899    Event Space  
2  Novello Theatre       51.512280        -0.119322        Theater  
3     The Delaunay       51.513181        -0.117988     Restaurant  
4        Lundenwic       51.512823        -0.118343    Coffee Shop

## Methodology

I will be comparing neighborhoods based on their attractions and a user's preferences. If a user enjoyed art, museums, and parks, I will use the data from Foursquare to find the areas that have the most options for those preferences. Then I can show the user which neighborhoods match up by showing the maps side by side.

Once we have the neighborhoods for both cities narrowed down by preference, we can create clusters using k-means clustering. I can then create side-by-side maps to show the neighborhoods that are similar which would allow a user to choose the best place to stay on their trip.

## Analysis

First, let's get a breakdown of the different venues in each neighborhood.

In [21]:
london_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                             
Aldwych                                            100   
Bankside                                           100   
Barbican                                           100   
Barnsbury                                           53   
Belgravia                                          100   
Blackfriars                                        100   
Bloomsbury                                         100   
Camden Town                                         94   
Charing Cross                                      100   
Chinatown                                          100   
Clerkenwell                                         38   
Covent Garden                                      100   
Elephant and Castle                                 85   
Farringdon                                         100   
Finsbury                                            38   
Fitzrovia                                          100   
Holborn                                            100   
Kennington                                         100   
King's Cross                                        99   
Knightsbridge                                      100   
Lambeth                                            100   
Marylebone (also St Marylebone)                    100   
Mayfair                                            100   
Millbank                                           100   
Newington                                          100   
Nine Elms                                           71   
Pentonville                                        100   
Pimlico                                            100   
Primrose Hill                                      100   
Soho                                               100   
Somerstown                                         100   
St Giles                                           100   
St James's                                         100   
St Luke's                                          100   
St Pancras                                         100   
Temple                                             100   
Vauxhall                                           100   
Walworth                                            62   
Westminster                                        100   

                                 Neighborhood Longitude  Venue  \
Neighborhood                                                     
Aldwych                                             100    100   
Bankside                                            100    100   
Barbican                                            100    100   
Barnsbury                                            53     53   
Belgravia                                           100    100   
Blackfriars                                         100    100   
Bloomsbury                                          100    100   
Camden Town                                          94     94   
Charing Cross                                       100    100   
Chinatown                                           100    100   
Clerkenwell                                          38     38   
Covent Garden                                       100    100   
Elephant and Castle                                  85     85   
Farringdon                                          100    100   
Finsbury                                             38     38   
Fitzrovia                                           100    100   
Holborn                                             100    100   
Kennington                                          100    100   
King's Cross                                         99     99   
Knightsbridge                                       100    100   
Lambeth                                             100    100   
Marylebone (also St Marylebone)                     100    100   
Mayfair                                             100    100   
Mill

In [22]:
paris_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                         
10th                            100                     100    100   
11th                            100                     100    100   
12th                             80                      80     80   
13th                            100                     100    100   
14th                            100                     100    100   
15th                            100                     100    100   
16th                             87                      87     87   
17th                            100                     100    100   
18th                            100                     100    100   
19th                            100                     100    100   
1st                             100                     100    100   
20th                             94                      94     94   
2nd                             100                     100    100   
3rd                             100                     100    100   
4th                             100                     100    100   
5th                             100                     100    100   
6th                             100                     100    100   
7th                             100                     100    100   
8th                             100                     100    100   
9th                             100                     100    100   

              Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                   
10th                     100              100             100  
11th                     100              100             100  
12th                      80               80              80  
13th                     100              100             100  
14th                     100              100             100  
15th                     100              100             100  
16th                      87               87              87  
17th                     100              100             100  
18th                     100              100             100  
19th                     100              100             100  
1st                      100              100             100  
20th                      94               94              94  
2nd                      100              100             100  
3rd                      100              100             100  
4th                      100              100             100  
5th                      100              100             100  
6th                      100              100             100  
7th                      100              100             100  
8th                      100              100             100  
9th                      100              100             100

In [23]:
print('There are {} unique categories.'.format(len(london_venues['Venue Category'].unique())))

There are 271 unique categories.


In [24]:
print('There are {} unique categories.'.format(len(paris_venues['Venue Category'].unique())))

There are 220 unique categories.


At this point, I want to combine the data into one dataframe so that I can cluster neighborhoods together across the two cities. I will need to get the top 10 types of venues in each neighborhood and then use k-means clustering.

In [25]:
combined_neighborhoods_venues = paris_venues.append(london_venues)

combined_neighborhoods_venues

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0             1st              48.864614                2.334396   
1             1st              48.864614                2.334396   
2             1st              48.864614                2.334396   
3             1st              48.864614                2.334396   
4             1st              48.864614                2.334396   
...           ...                    ...                     ...   
3635  Westminster              51.499615               -0.135236   
3636  Westminster              51.499615               -0.135236   
3637  Westminster              51.499615               -0.135236   
3638  Westminster              51.499615               -0.135236   
3639  Westminster              51.499615               -0.135236   

                          Venue  Venue Latitude  Venue Longitude  \
0     Musée des Arts Décoratifs       48.863077         2.333393   
1                     Ellsworth       48.865528         2.337057   
2           Place des Pyramides       48.863924         2.332224   
3                   Kosyuen 華修園       48.864163         2.333567   
4        Jardin du Palais Royal       48.864941         2.337728   
...                         ...             ...              ...   
3635                   Gianni's       51.494861        -0.127139   
3636       Market Hall Victoria       51.496378        -0.144286   
3637    The Goring Bar & Lounge       51.497385        -0.145475   
3638               Costa Coffee       51.500934        -0.124805   
3639  The Stafford London Hotel       51.506053        -0.139853   

         Venue Category  
0            Art Museum  
1     French Restaurant  
2                 Plaza  
3              Tea Room  
4                Garden  
...                 ...  
3635               Café  
3636         Food Court  
3637                Bar  
3638        Coffee Shop  
3639              Hotel  

[5601 rows x 7 columns]

In [26]:
print('There are {} unique categories.'.format(len(combined_neighborhoods_venues['Venue Category'].unique())))

There are 318 unique categories.


The number of restaurants and especially French restaurants in Paris have a large impact on the clustering, I have removed columns of restaurants to get a better idea of how neighborhoods are similar based on other attractions that the user would be more interested in while traveling. I have also removed the hotel column as that will not be needed.

In [27]:
combined_neighborhoods_onehot = pd.get_dummies(combined_neighborhoods_venues[['Venue Category']], prefix="", prefix_sep="")
combined_neighborhoods_onehot['Neighborhood'] = combined_neighborhoods_venues['Neighborhood']
cols = [c for c in combined_neighborhoods_onehot.columns if 'restaurant' not in c.lower() and 'hotel' not in c.lower()]

combined_neighborhoods_onehot = combined_neighborhoods_onehot[cols]

combined_neighborhoods_onehot.head()

Accessories Store  Animal Shelter  Antique Shop  Aquarium  Arcade  \
0                  0               0             0         0       0   
1                  0               0             0         0       0   
2                  0               0             0         0       0   
3                  0               0             0         0       0   
4                  0               0             0         0       0   

   Art Gallery  Art Museum  Arts & Crafts Store  Athletics & Sports  \
0            0           1                    0                   0   
1            0           0                    0                   0   
2            0           0                    0                   0   
3            0           0                    0                   0   
4            0           0                    0                   0   

   Auto Garage  ...  Warehouse Store  Watch Shop  Whisky Bar  Wine Bar  \
0            0  ...                0           0           0         0   
1            0  ...                0           0           0         0   
2            0  ...                0           0           0         0   
3            0  ...                0           0           0         0   
4            0  ...                0           0           0         0   

   Wine Shop  Wings Joint  Women's Store  Yoga Studio  Zoo  Zoo Exhibit  
0          0            0              0            0    0            0  
1          0            0              0            0    0            0  
2          0            0              0            0    0            0  
3          0            0              0            0    0            0  
4          0            0              0            0    0            0  

[5 rows x 237 columns]

In [28]:
combined_grouped = combined_neighborhoods_onehot.groupby('Neighborhood').mean().reset_index()
combined_grouped

Neighborhood  Accessories Store  Animal Shelter  \
0                              10th               0.00        0.000000   
1                              11th               0.00        0.000000   
2                              12th               0.00        0.000000   
3                              13th               0.00        0.000000   
4                              14th               0.00        0.000000   
5                              15th               0.00        0.000000   
6                              16th               0.00        0.000000   
7                              17th               0.00        0.000000   
8                              18th               0.00        0.000000   
9                              19th               0.00        0.000000   
10                              1st               0.00        0.000000   
11                             20th               0.00        0.000000   
12                              2nd               0.00        0.000000   
13                              3rd               0.00        0.000000   
14                              4th               0.00        0.000000   
15                              5th               0.00        0.000000   
16                              6th               0.00        0.000000   
17                              7th               0.00        0.000000   
18                              8th               0.00        0.000000   
19                              9th               0.00        0.000000   
20                          Aldwych               0.00        0.000000   
21                         Bankside               0.00        0.000000   
22                         Barbican               0.00        0.000000   
23                        Barnsbury               0.00        0.000000   
24                        Belgravia               0.00        0.000000   
25                      Blackfriars               0.00        0.000000   
26                       Bloomsbury               0.01        0.000000   
27                      Camden Town               0.00        0.000000   
28                    Charing Cross               0.00        0.000000   
29                        Chinatown               0.00        0.000000   
30                      Clerkenwell               0.00        0.000000   
31                    Covent Garden               0.00        0.000000   
32              Elephant and Castle               0.00        0.000000   
33                       Farringdon               0.00        0.000000   
34                         Finsbury               0.00        0.000000   
35                        Fitzrovia               0.00        0.000000   
36                          Holborn               0.01        0.000000   
37                       Kennington               0.00        0.000000   
38                     King's Cross               0.00        0.000000   
39                    Knightsbridge               0.00        0.000000   
40                          Lambeth               0.00        0.000000   
41  Marylebone (also St Marylebone)               0.00        0.000000   
42                          Mayfair               0.00        0.000000   
43                         Millbank               0.00        0.000000   
44                        Newington               0.00        0.000000   
45                        Nine Elms               0.00        0.014085   
46                      Pentonville               0.00        0.000000   
47                          Pimlico               0.00        0.000000   
48                    Primrose Hill               0.00        0.000000   
49                             Soho               0.00        0.000000   
50                       Somerstown               0.00        0.000000   
51                         St Giles               0.01        0.000000   
52                       St James's               0.00        0.000000   
53                        S

In [29]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [30]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = combined_grouped['Neighborhood']

for ind in np.arange(combined_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(combined_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                              10th           Coffee Shop   
1                              11th                   Bar   
2                              12th                Bistro   
3                              13th                Bakery   
4                              14th                   Bar   
5                              15th                Bistro   
6                              16th                Bakery   
7                              17th              Wine Bar   
8                              18th                   Bar   
9                              19th                   Bar   
10                              1st                 Plaza   
11                             20th                   Bar   
12                              2nd              Wine Bar   
13                              3rd           Art Gallery   
14                              4th                 Plaza   
15                              5th                 Plaza   
16                              6th                 Plaza   
17                              7th                 Plaza   
18                              8th                Garden   
19                              9th                Bakery   
20                          Aldwych               Theater   
21                         Bankside                   Pub   
22                         Barbican  Gym / Fitness Center   
23                        Barnsbury         Grocery Store   
24                        Belgravia              Boutique   
25                      Blackfriars           Coffee Shop   
26                       Bloomsbury           Coffee Shop   
27                      Camden Town                   Pub   
28                    Charing Cross               Theater   
29                        Chinatown                Bakery   
30                      Clerkenwell           Coffee Shop   
31                    Covent Garden          Burger Joint   
32              Elephant and Castle                   Pub   
33                       Farringdon           Coffee Shop   
34                         Finsbury           Coffee Shop   
35                        Fitzrovia           Coffee Shop   
36                          Holborn               Theater   
37                       Kennington                  Café   
38                     King's Cross                  Café   
39                    Knightsbridge                  Café   
40                          Lambeth                  Café   
41  Marylebone (also St Marylebone)           Coffee Shop   
42                          Mayfair              Boutique   
43                         Millbank           Coffee Shop   
44                        Newington                   Pub   
45                        Nine Elms           Pizza Place   
46                      Pentonville                   Pub   
47                          Pimlico                  Café   
48                    Primrose Hill           Zoo Exhibit   
49                             Soho           Coffee Shop   
50                       Somerstown           Coffee Shop   
51                         St Giles               Theater   
52                       St James's              Boutique   
53                        St Luke's           Coffee Shop   
54                       St Pancras           Coffee Shop   
55                           Temple               Theater   
56                         Vauxhall                  Café   
57                         Walworth                   Pub   
58                      Westminster           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Pizza Place        Breakfast Spot                Bakery   
1           Cocktail Bar                Bistro           Pizza Place   
2            Supermarket                Bakery              Bus Stop   
3               Creperie                Bistro             Brasserie   
4 

In [31]:
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

In [32]:
kclusters = 5

combined_grouped_clustering = combined_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(combined_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:] 

array([4, 2, 2, 0, 2, 0, 0, 0, 2, 2, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 3, 1,
       4, 1, 0, 4, 4, 1, 3, 0, 4, 3, 1, 4, 4, 4, 3, 1, 4, 0, 1, 4, 0, 4,
       1, 1, 4, 4, 1, 4, 4, 3, 0, 4, 4, 3, 1, 1, 4], dtype=int32)

In [33]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

neighborhoods_venues_sorted

Cluster Labels                     Neighborhood 1st Most Common Venue  \
0                4                             10th           Coffee Shop   
1                2                             11th                   Bar   
2                2                             12th                Bistro   
3                0                             13th                Bakery   
4                2                             14th                   Bar   
5                0                             15th                Bistro   
6                0                             16th                Bakery   
7                0                             17th              Wine Bar   
8                2                             18th                   Bar   
9                2                             19th                   Bar   
10               0                              1st                 Plaza   
11               2                             20th                   Bar   
12               0                              2nd              Wine Bar   
13               0                              3rd           Art Gallery   
14               0                              4th                 Plaza   
15               0                              5th                 Plaza   
16               0                              6th                 Plaza   
17               0                              7th                 Plaza   
18               0                              8th                Garden   
19               0                              9th                Bakery   
20               3                          Aldwych               Theater   
21               1                         Bankside                   Pub   
22               4                         Barbican  Gym / Fitness Center   
23               1                        Barnsbury         Grocery Store   
24               0                        Belgravia              Boutique   
25               4                      Blackfriars           Coffee Shop   
26               4                       Bloomsbury           Coffee Shop   
27               1                      Camden Town                   Pub   
28               3                    Charing Cross               Theater   
29               0                        Chinatown                Bakery   
30               4                      Clerkenwell           Coffee Shop   
31               3                    Covent Garden          Burger Joint   
32               1              Elephant and Castle                   Pub   
33               4                       Farringdon           Coffee Shop   
34               4                         Finsbury           Coffee Shop   
35               4                        Fitzrovia           Coffee Shop   
36               3                          Holborn               Theater   
37               1                       Kennington                  Café   
38               4                     King's Cross                  Café   
39               0                    Knightsbridge                  Café   
40               1                          Lambeth                  Café   
41               4  Marylebone (also St Marylebone)           Coffee Shop   
42               0                          Mayfair              Boutique   
43               4                         Millbank           Coffee Shop   
44               1                        Newington                   Pub   
45               1                        Nine Elms           Pizza Place   
46               4                      Pentonville                   Pub   
47               4                          Pimlico                  Café   
48               1                    Primrose Hill           Zoo Exhibit   
49               4                             Soho           Coffee Shop   
50               4                       Somerstown           Coffee Shop   


In [34]:
london_merged = london_city_center_df

london_merged.rename(columns= {'Location': 'Neighborhood'}, inplace=True)

london_venues_sorted = neighborhoods_venues_sorted[20:]
london_merged = london_merged.join(london_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

london_merged

Neighborhood      London borough Post town  \
0                           Aldwych    Westminster [10]    LONDON   
1                          Bankside      Southwark [14]    LONDON   
2                          Barbican           City [14]    LONDON   
3                         Barnsbury      Islington [17]    LONDON   
4                         Belgravia    Westminster [22]    LONDON   
5                       Blackfriars           City [27]    LONDON   
6                        Bloomsbury         Camden [29]    LONDON   
7                       Camden Town         Camden [40]    LONDON   
8                     Charing Cross         Westminster    LONDON   
9                         Chinatown         Westminster    LONDON   
10                      Clerkenwell           Islington    LONDON   
11                    Covent Garden         Westminster    LONDON   
12              Elephant and Castle           Southwark    LONDON   
13                       Farringdon     Islington, City    LONDON   
14                         Finsbury           Islington    LONDON   
15                        Fitzrovia              Camden    LONDON   
16                          Holborn              Camden    LONDON   
17                       Kennington  Lambeth, Southwark    LONDON   
18                     King's Cross   Camden, Islington    LONDON   
19                    Knightsbridge         Westminster    LONDON   
20                          Lambeth             Lambeth    LONDON   
21  Marylebone (also St Marylebone)         Westminster    LONDON   
22                          Mayfair         Westminster    LONDON   
23                         Millbank         Westminster    LONDON   
24                        Newington           Southwark    LONDON   
25                        Nine Elms          Wandsworth    LONDON   
26                      Pentonville           Islington    LONDON   
27                          Pimlico         Westminster    LONDON   
28                    Primrose Hill              Camden    LONDON   
29                             Soho         Westminster    LONDON   
30                       Somerstown              Camden    LONDON   
31                       St James's         Westminster    LONDON   
32                         St Giles              Camden    LONDON   
33                        St Luke's           Islington    LONDON   
34                       St Pancras              Camden    LONDON   
35                           Temple   City, Westminster    LONDON   
36                         Vauxhall             Lambeth    LONDON   
37                         Walworth           Southwark    LONDON   
38                      Westminster         Westminster    LONDON   

   Postcode district OS grid ref   Latitude  Longitude City Center  \
0                WC2    TQ307810  51.512819  -0.117388        True   
1                SE1    TQ325795  51.498921  -0.092006        True   
2           EC1, EC2    TQ322818  51.519660  -0.095466        True   
3                 N1    TQ305845  51.544318  -0.118974        True   
4                SW1    TQ283792  51.497193  -0.152637        True   
5                EC4    TQ318808  51.510767  -0.101607        True   
6                WC1    TQ299820  51.521990  -0.128550        True   
7                NW1    TQ295845  51.544548  -0.133398        True   
8                WC2    TQ305805  51.508372  -0.120456        True   
9                 W1    TQ297808  51.511252  -0.131875        True   
10               EC1    TQ315825  51.526113  -0.105296        True   
11               WC2    TQ303809  51.512013  -0.123190        True   
12   SE1, SE11, SE17    TQ319789  51.493669  -0.100876        True   
13               EC1    TQ315818  51.519823  -0.105558        True   
14               EC1    TQ315825  51.526113  -0.105296        True   
15                W1    TQ293816  51.518533  -0.137347        True   
16          WC1, WC2    TQ305815  51.517359  -0.120085        True   
17

In [35]:
london_clusters = folium.Map(location=[london_latitude, london_longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_merged['Latitude'], london_merged['Longitude'], london_merged['Neighborhood'], london_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(london_clusters)

In [36]:
paris_merged = paris_df

paris_merged.rename(columns= {'Arrondissement': 'Neighborhood'}, inplace=True)

paris_venues_sorted = neighborhoods_venues_sorted[:20]
paris_merged = paris_merged.join(paris_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

paris_merged

Neighborhood   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0           1st  48.864614   2.334396               0                 Plaza   
1           2nd  48.868743   2.341688               0              Wine Bar   
2           3rd  48.864212   2.360936               0           Art Gallery   
3           4th  48.856202   2.355619               0                 Plaza   
4           5th  48.845973   2.344350               0                 Plaza   
5           6th  48.850433   2.332951               0                 Plaza   
6           7th  48.857028   2.320195               0                 Plaza   
7           8th  48.870905   2.312152               0                Garden   
8           9th  48.876019   2.339962               0                Bakery   
9          10th  48.876126   2.359839               4           Coffee Shop   
10         11th  48.858416   2.379703               2                   Bar   
11         12th  48.839615   2.395752               2                Bistro   
12         13th  48.832397   2.355583               0                Bakery   
13         14th  48.833059   2.326959               2                   Bar   
14         15th  48.841370   2.300383               0                Bistro   
15         16th  48.863171   2.275765               0                Bakery   
16         17th  48.884224   2.322364               0              Wine Bar   
17         18th  48.890012   2.346467               2                   Bar   
18         19th  48.889343   2.384360               2                   Bar   
19         20th  48.865042   2.398929               2                   Bar   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Café              Wine Bar            Art Museum   
1                 Bakery      Pedestrian Plaza              Creperie   
2           Cocktail Bar           Coffee Shop              Wine Bar   
3                 Garden                Bistro           Coffee Shop   
4                 Bakery                   Bar             Bookstore   
5               Wine Bar        Chocolate Shop          Cocktail Bar   
6             Art Museum                Bakery                Garden   
7                  Plaza        Clothing Store              Boutique   
8            Pizza Place              Wine Bar                  Café   
9            Pizza Place        Breakfast Spot                Bakery   
10          Cocktail Bar                Bistro           Pizza Place   
11           Supermarket                Bakery              Bus Stop   
12              Creperie                Bistro             Brasserie   
13                Bakery                Bistro               Theater   
14                  Park                 Plaza                Bakery   
15           Supermarket        Sandwich Place                  Café   
16        Farmers Market                   Bar           Pastry Shop   
17                Bistro                 Plaza           Pizza Place   
18           Supermarket                  Café                Bistro   
19                Bistro                 Plaza                Bakery   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Theater         Historic Site                Garden   
1            Coffee Shop             Bookstore          Cocktail Bar   
2         Sandwich Place          Gourmet Shop        Clothing Store   
3           Gourmet Shop        Clothing Store            Art Museum   
4    Indie Movie Theater           Coffee Shop              Creperie   
5            Pastry Shop              Fountain            Shoe Store   
6                   Café         Historic Site        History Museum   
7            Tailor Shop           Coffee Shop          Cocktail Bar   
8           Cocktail Bar              Creperie           Cheese Shop   
9           Cocktail Bar             Wine Shop                Bistro   
10                  Café           Rec

In [37]:
paris_clusters = folium.Map(location=[paris_latitude, paris_longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(paris_merged['Latitude'], paris_merged['Longitude'], paris_merged['Neighborhood'], paris_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(paris_clusters)

In [38]:
from IPython.core.display import display, HTML

htmlmap = HTML('<iframe srcdoc="{}" style="float:left; width: {}px; height: {}px; display:inline-block; width: 50%; margin: 0 auto; border: 2px solid black"></iframe>'
           '<iframe srcdoc="{}" style="float:right; width: {}px; height: {}px; display:inline-block; width: 50%; margin: 0 auto; border: 2px solid black"></iframe>'
           .format(london_clusters.get_root().render().replace('"', '&quot;'),500,500,
                   paris_clusters.get_root().render().replace('"', '&quot;'),500,500))
display(htmlmap)

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/IPython/core/display.py:717: UserWarning: Consider using IPython.display.IFrame instead
  warnings.warn("Consider using IPython.display.IFrame instead")


## Results and Discussion

In the end, we can see that there are some clear differences between Paris and London. London does not have any neighborhoods in Cluster 2 which is mainly bakeries and bistros (which makes sense for Paris),  Paris has no neighborhoods in Cluster 1 which has many pubs (which makes sense for London) or Cluster 3 which has many theatres (which also makes sense with the West End in London).

The neighborhoods that are similar seem to be broken down into areas that have plazas and parks/gardens (Cluster 0) and coffee shops and cafes (Cluster 4).

A user could make a decision on what area to stay in based on the clusters provided in both maps. Even with the different types of clusters in London, they can at least see where those types of attractions are. That means that they may not have found and liked many pubs in Paris the year before, but they know where they can stay that has those in London based on the map. If they want to stick to what they know from Paris, they can find those two clusters that are similar and go with that.

While the results are reasonable, further data gathering could be useful to fine-tune the clusters. Also, cleaning the data more to just include major attractions instead of all the different types of venues that Foursquare offers might make it better for tourists looking more for the traditional experience instead of seeing pubs, cafes, and plazas are.

## Conclusion

The purpose of the project was to be able to make recommendations to a user on where to stay in London based on their experiences in Paris the year before. By limiting the regions to search and the types of venues to look for,  we have been able to determine clusters of neighborhoods that have similar or dissimilar features for the user to use to determine where they want to stay. The user will be able to make the final decision on where to stay based on the side-by-side comparison and their preferences.